rhombus-lib/rhombus/private/amalgam/filesystem.rhm

```diff
@@ -47,5 +71,344 @@ namespace filesystem:
+    cond
+    | !recur:
+        if !follow_links && link_exists(src_path)
+        | copy_link(src_path, dest_path)
+        | copy_file(src_path, dest_path)
+    | ~else:
+        fun recur(src_path, dest_path):
+          fun copy_dir():
+            when !exists_ok || !directory_exists(dest_path)
+            | rkt_error.local_error_adjust { #'#{make-directory}: who }:
+                rkt.#{make-directory}(dest_path, permissions || 0o777)
+            let files :~ PairList:
+              rkt_error.local_error_adjust { #'#{directory-list}: who }:
+                rkt.#{directory-list}(src_path)
+            for (file: files):
+              recur(Path.add(src_path, file), Path.add(dest_path, file))
+          match (rkt_error.local_error_adjust { #'#{file-or-directory-type}: who }:
+                   rkt.#{file-or-directory-type}(src_path))
+          | #'file: copy_file(src_path, dest_path)
+          | #'link:
+              if !follow_links
+              | copy_link(src_path, dest_path)
+              | copy_file(src_path, dest_path)
+          | #'#{directory-link}:
+              if !follow_links
+              | copy_link(src_path, dest_path)
+              | copy_dir()
+          | #'directory:
+              copy_dir()
+          | ~else:
+              not_a_path_error(who, src_path)
+        recur(src_path, dest_path)
+
+  fun size(path :: PathString) :: NonnegInt:
+    ~who: who
+    rkt_error.local_error_adjust { #'#{file-size}: who }:
+      rkt.#{file-size}(path)
+
+  fun default_fail(): error("should not get here")
+
+  fun modify_seconds(path :: PathString,
+                     ~change_to: change_to :: maybe(Int) = #false,
+                     ~fail: fail_k :: Function.of_arity(0) = default_fail):
+    ~who: who
+    rkt_error.local_error_adjust { #'#{file-or-directory-modify-seconds}: who }:
+      if fail_k == default_fail
+      | rkt.#{file-or-directory-modify-seconds}(path, change_to)
+      | rkt.#{file-or-directory-modify-seconds}(path, change_to, fail_k)
+
+  fun permissions(path :: PathString,
+                  ~bits = #false,
+                  ~change_to: change_to :: maybe(Int.in(0, 65535)) = #false):
+    ~who: who
+    rkt_error.local_error_adjust { #'#{file-or-directory-permissions}: who }:
+      cond
+      | change_to:
+          rkt.#{file-or-directory-permissions}(path, change_to)
+      | bits:
+          List(& rkt.#{file-or-directory-permissions}(path, #'bits_to))
+      | ~else:
+          List(& rkt.#{file-or-directory-permissions}(path, #false))
+
+  fun identity(path :: PathString,
+               ~follow_links = #true) :: PosInt:
+    ~who: who
+    rkt_error.local_error_adjust { #'#{file-or-directory-identity}: who }:
+      rkt.#{file-or-directory-identity}(path, #{#:as-link?}: !follow_links)
+
+  fun stat(path :: PathString,
+           ~follow_links = #true) :~ Map:
+    ~who: who
+    rkt_error.local_error_adjust { #'#{file-or-directory-stat}: who }:
```

Comment:
The keys in the returned hash should probably not use hyphens.

Reply:
It's currently documented as "low level, so using Racket symbol conventions". My thought here was that creating a new map to rename keys is a lot of work with little payoff, and using the function is already a much leakier abstraction than the others.